lib/class-wp-rest-plugins-controller.php

```diff
@@ -0,0 +1,822 @@
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function create_item( $request ) {
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+		require_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+		require_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+
+		$slug = $request['slug'];
+
+		// Verify filesystem is accessible first.
+		$filesystem_available = $this->is_filesystem_available();
+		if ( is_wp_error( $filesystem_available ) ) {
+			return $filesystem_available;
+		}
+
+		$api = plugins_api(
+			'plugin_information',
+			array(
+				'slug'   => $slug,
+				'fields' => array(
+					'sections' => false,
+				),
+			)
+		);
+
+		if ( is_wp_error( $api ) ) {
+			if ( false !== strpos( $api->get_error_message(), 'Plugin not found.' ) ) {
+				$api->add_data( array( 'status' => 404 ) );
+			} else {
+				$api->add_data( array( 'status' => 500 ) );
+			}
+
+			return $api;
+		}
+
+		$skin     = new WP_Ajax_Upgrader_Skin();
+		$upgrader = new Plugin_Upgrader( $skin );
+
+		$result = $upgrader->install( $api->download_link );
+
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+
+		// This should be the same as $result above.
+		if ( is_wp_error( $skin->result ) ) {
+			return $skin->result;
+		}
+
+		if ( $skin->get_errors()->has_errors() ) {
+			return $skin->get_errors();
+		}
+
+		if ( is_null( $result ) ) {
+			global $wp_filesystem;
+			// Pass through the error from WP_Filesystem if one was raised.
+			if ( $wp_filesystem instanceof WP_Filesystem_Base && is_wp_error( $wp_filesystem->errors ) && $wp_filesystem->errors->has_errors() ) {
+				return new WP_Error( 'unable_to_connect_to_filesystem', $wp_filesystem->errors->get_error_message(), array( 'status' => 500 ) );
+			}
+
+			return new WP_Error( 'unable_to_connect_to_filesystem', __( 'Unable to connect to the filesystem. Please confirm your credentials.', 'gutenberg' ), array( 'status' => 500 ) );
+		}
+
+		$file = $upgrader->plugin_info();
+
+		if ( ! $file ) {
+			return new WP_Error( 'unable_to_determine_installed_plugin', __( 'Unable to determine what plugin was installed.', 'gutenberg' ), array( 'status' => 500 ) );
+		}
+
+		if ( 'inactive' !== $request['status'] ) {
+			$can_change_status = $this->plugin_status_permission_check( $file, $request['status'], 'inactive' );
+
+			if ( is_wp_error( $can_change_status ) ) {
+				return $can_change_status;
+			}
+
+			$changed_status = $this->handle_plugin_status( $file, $request['status'], 'inactive' );
+
+			if ( is_wp_error( $changed_status ) ) {
+				return $changed_status;
+			}
+		}
+
+		$path          = WP_PLUGIN_DIR . '/' . $file;
+		$data          = get_plugin_data( $path, false, false );
+		$data['_file'] = $file;
+
+		$response = $this->prepare_item_for_response( $data, $request );
+		$response->set_status( 201 );
```

Comment:
301?

Reply:
`201` is correct for a created response.